hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodieFileGroupReader.java

```diff
@@ -146,28 +154,52 @@ public void initRecordIterators() {
    * @return {@code true} if the next record exists; {@code false} otherwise.
    * @throws IOException on reader error.
    */
-  public boolean hasNext() throws IOException {
+  public boolean hasNext() {
+    // Merge records from base file and log files.
+    int baseFileSequenceNo = 0;
     while (baseFileIterator.hasNext()) {
       T baseRecord = baseFileIterator.next();
-      String recordKey = readerContext.getRecordKey(baseRecord, readerState.baseFileAvroSchema);
-      Pair<Option<T>, Map<String, Object>> logRecordInfo = logFileRecordMapping.remove(recordKey);
-      Option<T> resultRecord = logRecordInfo != null
-          ? merge(Option.of(baseRecord), Collections.emptyMap(), logRecordInfo.getLeft(), logRecordInfo.getRight())
-          : merge(Option.empty(), Collections.emptyMap(), Option.of(baseRecord), Collections.emptyMap());
+      Pair<Option<T>, Map<String, Object>> logRecordInfo;
+
+      if (shouldUseRecordPosition) {
```

Comment:
Kind of feel we should use extend instead of messing up the very basic reader with ramified logic, like what we do to the cdc writer, we have `HoodieMergeHandle` and also `HoodieMergeHandleWithChangeLog`.

Reply:
Same applies to `BaseHoodieLogRecordReader`

Reply:
Good point. I also feel if-else is not very clean for new logic. After adding partial-merging, we will have many readers.